spec/src/main/asciidoc/microprofile-lra-spec.adoc

```diff
@@ -555,38 +555,154 @@ of these methods. If it knows it will never be able to compensate or complete th
-a JAX-RS `@PUT` annotation the error SHOULD be reported using a JAX-RS
+The participant marker annotations are annotations that allow users to mark a method 
+for the execution by the LRA implemenatation according to the <<participant-state-model,
+the participant state model>>. These annotations are: 
+
+* `@Compensate` -- a method to be executed when the LRA is cancelled
+* `@Complete` -- a method to be executed when the LRA is completed
+* `@Status` -- a method that allow user to state status of the participant with regards 
+to a paricular LRA
+* `@Forget` -- a method to be executed when the LRA allows participant to 
+clear all associated information
+
+This specification is generally expected to be associated with the JAX-RS runtime. However, 
+we acknowledge that the mapping to JAX-RS resources may not always be feasible which is 
+why we distinguish two types of participant method definitions -- associated with the 
+JAX-RS resource method or not bound to JAX-RS. 
+
+[[jaxrs-participant-methods]]
+===== JAX-RS participant methods
+
+The following table presents expectations that are placed on individual participant 
+annotations when associated with JAX-RS resource methods:
+
+[[jaxrs-response-table]]
+|===
+| Annotation | HTTP method | Expected status codes | Response
+
+| `@Compensate`
+| PUT
+| 200, 202, 404, 412
+| <<source-ParticipantStatus,ParticipantStatus>>
+
+| `@Complete`
+| PUT
+| 200, 202, 404, 412
+| <<source-ParticipantStatus,ParticipantStatus>>
+
+| `@Status`
+| GET
+| 200, 202, 404, 412
+| <<source-ParticipantStatus,ParticipantStatus>>
+
+| `@Forget`
+| DELETE
+| 200, 404, 412
+| no expectations
+
+|=== 
+
+If the participant annotation is not accompanied by an associated JAX-RS HTTP method 
+annotation according to the <<jaxrs-response-table,table above>>, the error SHOULD be reported using a JAX-RS 
 exception mapper that maps to a `412 Precondition Failed` HTTP status code.
 
+[[non-jaxrs-participant-methods]]
+===== Non-JAX-RS participant methods
+
+When the participant annotations are applied to the non-JAX-RS resource methods they 
+MUST adhere to these predefined signatures:
+
+* *Return type*: 
+** `void`: successfull execution is mapped to `Compensated` or `Completed` participant statuses, 
+ error execution is handled by exception mapping
+*** applicable only for `@Compensate` and `@Complete` participant methods
+** <<source-ParticipantStatus,ParticipantStatus>>
+** `javax.ws.rs.core.Response`: handled similarly as for 
+<<jaxrs-participant-methods, JAX-RS participant methods>>
+* *Arguments*: up to 2 arguments of types:
+** `java.net.URI`: representing LRA id
+** `java.lang.String`: representing user data
+
+Declaring more than two arguments, different types of arguments or different return type 
+for any non-JAX-RS method annotatated with the participant marker annotation MUST result 
+in the deployment time `InvalidLRAParticipantDefinitionException`. 
+Please note that both arguments are optional.
```

Comment:
I've modeled it after JAX-RS - `@HeaderParam` is optional so users don't have to use lraId

Reply:
The spec says:

> When a method is invoked in the context of an LRA its identifier MUST be made available to the
business logic

Reply:
it is in both JAX-RS and non-JAX-RS and it's up to the user if he/she defines it in signature as header / parameter and use it. But of course it can be required but I just want to stay aligned as much as possible with the way how we do things in JAX-RS so you can just add / delete a few JAX-RS annotations and participant methods should still work. In other words, this should be valid compensate method right now:

```
@PUT
@Path("compensate")
@Compensate
public Response compensate() {
      return Response.ok().build();
}
```

Reply:
if user defines URI parameter he/she will get the lraId in it. Is that sufficient?

Reply:
So is this something Rudy asked for. Can you provide an example/explanation of how a participant compensates without the context?

Reply:
If you can provide an example then that is fine but if we cannot come up with one then it would make sense to make the parameter mandatory, @rdebusscher can you think of an example where the LRA id is not required to perform the compensation

Reply:
@mmusgrov I think you always will use  the LRA id for the compensate (and for complete)

But I agree with @xstefank, the spec says that the implementation must make the LRA id available to the business logic. But it not mandates that you specify a method parameter annotated with @HeaderParam. Some from the spec/implementation side it MUST be available but business logic can decide if it requires or not.

So if we follow the same line of thoughts, a method parameter is not needed here also.

Reply:
@mmusgrov If we say the businessLogic MUST use (or at least accept) the LRA id, we must indicate that in the spec + require that the implementation verifies this requirement at deployment time. So it needs to scan all methods annotated wit @Complete and @Compensate and verify it has at least a parameter annotated with @HeaderParam("Long-Running-Action").

And if it not complies with it, fail the deployment.

And of course a TCK to verify this situation.

Reply:
>I know It is not possible to force applications to explicitly receive a particular header but if we could we have done so.

Why? At deployment we can check to see if the @HeaderParam("Long-Running-Action") is specified as method parameter!

Reply:
That's a good point. Mind you, I was thinking on the way home that any service that always compensates in the same way would not need the id. For example a warehousing application which registers a compensation that just triggers a general stock check taking appropriate actions on items which are low.

Reply:
@mmusgrov exactly. There may also be stack like use-cases where `@LRA` will do +1 and `@Compensate` will do -1.

Please remember that `@HeaderParam` is not the only way how users can obtain JAX-RS headers.

Personally I would like to avoid more scanning on deployment if it's not necessary. The specification is rightfully stating that the LRAId must be available for the user but what we are discussing here is whether the user makes use of it. I think it's natural to expect that when someone will be writing participant method he/she will add lraId if there will be need for it.